Sources/FoundationEssentials/String/String+Internals.swift

```diff
@@ -35,15 +40,40 @@ extension String {
         // leading slash indicates a rooted path on the drive for the current
         // working directory.
         return try Substring(self.utf8.dropFirst(bLeadingSlash ? 1 : 0)).withCString(encodedAs: UTF16.self) { pwszPath in
+            // lt MAX_PATH not lte MAX_PATH because MAX_PATH includes the null terminator
+            if preserveRelativePaths && wcslen(pwszPath) < MAX_PATH {
+                return try body(pwszPath)
+            }
+
             // 1. Normalize the path first.
+            // Contrary to the documentation, this works on long paths independently
+            // of the registry or process setting to enable long paths (but it will also
+            // not add the \\?\ prefix required by other functions under these conditions).
             let dwLength: DWORD = GetFullPathNameW(pwszPath, 0, nil, nil)
-            return try withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwLength)) {
-                guard GetFullPathNameW(pwszPath, DWORD($0.count), $0.baseAddress, nil) > 0 else {
+            return try withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwLength)) { fullPath in
+                guard GetFullPathNameW(pwszPath, DWORD(fullPath.count), fullPath.baseAddress, nil) > 0 else {
                     throw CocoaError.errorWithFilePath(self, win32: GetLastError(), reading: true)
                 }
 
-                // 2. Perform the operation on the normalized path.
-                return try body($0.baseAddress!)
+                // 1.5 Leave \\.\ prefixed paths alone since device paths are already an exact representation and PathCchCanonicalizeEx will mangle these.
+                if let base = fullPath.baseAddress, base[0] == UInt8(ascii: "\\"), base[1] == UInt8(ascii: "\\"), base[2] == UInt8(ascii: "."), base[3] == UInt8(ascii: "\\") {
+                    return try body(base)
+                }
+
+                // 2. Canonicalize the path.
+                // This will add the \\?\ prefix if needed based on the path's length.
+                let capacity = Int16.max
+                return try withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: numericCast(capacity)) { outBuffer in
+                    let flags: ULONG = PATHCCH_ALLOW_LONG_PATHS
+                    let result = PathCchCanonicalizeEx(outBuffer.baseAddress, numericCast(capacity), fullPath.baseAddress, flags)
```

Comment:
Rather than the `switch` on the result, why not `guard` since we only consider one case?

Reply:
Done, I also changed this to call PathAllocCanonicalize which could theoretically be more efficient by allocating a smaller buffer.